### PR TITLE
Draft: Add tools api support

### PR DESCRIPTION
**Pull request by jrmi**
This MR is not ready for review yet, but I have a first working version with a limited set of tools. Let me know what you think.

### Old comment

What I've done so far:

- Added a new CLI flag (-f or --function-call) to activate tools. I marked this option as experimental, and I believe it should remain experimental for a while.
- Ported 4 tools to the function calling. I want to address the questions before proceeding further.
- In function calling mode, the tools are no longer prompted (but something replaces it, see below).

We need a new description for the tool that differs from the existing one, as the tool works differently. We also need a new execute function, as the parameter system is different. Even the examples are no longer working since they use the current syntax.

I think we have two options:

- Stick with the same objects and simply add the new properties (better naming suggestions are welcome :smiley:). This is what I started to implement.
- Create a new class dedicated to the new objects and register these instead of the existing ones when using the function calling feature. I think it also makes sense because we hardly reuse anything from the prompted tools.

I'm unsure what we should do with the "aliases" tools, those that don't have an execute function because they're just an "alias" to another tool, like the read function. I proposed a separate description for these tools as well, and in the initial prompt, they are still included in a section that gives the model some special use cases for these tools. These tools are no longer registered, otherwise, the model attempts to call them. I think we could get rid of these if we provided a more detailed description of the aliased tools, but I might be wrong.

On a side note, I tried using the `lama3.2:1B` model (which is a small model) with function calling, and it was able to call a tool, which wasn't the case with the prompted tools. However, OpenAI doesn't detect it as a tool call, so the raw JSON is returned. We could parse it to get an actual tool call, but that is something for later.

I've added the `json_repair` dependency to parse the parameter JSON because there is a high chance that the model generates something that isn't parsable, as it doesn't protect necessary characters. This dependency was created for this purpose.

@ErikBjare, I'd love to have your thoughts on this before I continue.

----

> [!IMPORTANT]
> Adds tools API support with function call execution, updating chat loop, CLI, and tool specifications.
> 
>   - **Behavior**:
>     - Adds `use_function_call` parameter to `chat()` and `step()` in `chat.py` to enable function call usage.
>     - Updates `main()` in `cli.py` to include `--tool-format` option for CLI.
>     - Modifies `reply()` in `llm/__init__.py` to handle tools in responses.
>   - **Tool Integration**:
>     - Updates `ToolSpec` in `tools/base.py` to include `execute_fn` for function call execution.
>     - Implements `APIToolUse` class in `tools/base.py` for handling API tool calls.
>     - Updates `tool_save` and `tool_append` in `tools/save.py` to support function calls.
>     - Updates `tool` in `tools/shell.py` to support function calls.
>   - **Misc**:
>     - Adds `json-repair` dependency in `pyproject.toml` for JSON handling.
>     - Updates `prompt_tools()` in `prompts.py` to handle function call descriptions.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=ErikBjare%2Fgptme&utm_source=github&utm_medium=referral)<sup> for 66e9f6209e65091ade2464105a32e843a4cca8e5. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

File: gptme/cli.py
Comment (jrmi) [reply gptme/cli.py]: Fixed.

File: gptme/tools/base.py
Comment (ErikBjare) [inline gptme/tools/base.py]: I think we should somehow merge this with the already existing `instructions` property.

In the cases where `instructions` contain format-specific instructions, we could move such instructions into format-specific prompting under a new `instructions_format` dict with keys like `markdown`.

File: gptme/tools/base.py
Comment (ErikBjare) [inline gptme/tools/base.py]: I think we should be able to solve this without subclassing. We could support both `args` (list) and `kwargs` (dict) in the same ToolUse class.

File: gptme/tools/base.py
Comment (ErikBjare) [inline gptme/tools/base.py]: This is confusingly similarly named, we should find a way to reuse the `ExecuteFunc` protocol, probably by using it and mapping the parameters.

File: gptme/tools/save.py
Comment (ErikBjare) [inline gptme/tools/save.py]: As the comment suggests on line 15 above, we should refactor this to not mention markdown-specific instructions in the mail instructions, and move them into a separate `instructions_format` dict in the ToolSpec.

File: gptme/tools/base.py
Comment (ErikBjare) [inline gptme/tools/base.py]:
```suggestion
                    content = self.args.get("content")
                    path = self.args.get("filepath") or self.args.get("path")
                    ex = tool.execute(content, [path] if path else [], confirm)
```

File: gptme/tools/base.py
Comment (ErikBjare) [inline gptme/tools/base.py]: I really don't think we need to subclass for this.

**Review by ErikBjare — COMMENTED** [review]
Good stuff!

**Comment by jrmi** [comment]
> Good stuff!

Thanks for the first review. Everything makes sense and I tried to go in that direction as much as possible. As bonus you I have now implemented the `--tool-format` that can have the value `markdown`, `xml`, `tool` to change the tool format. 

I removed all the extra subclasses and there is now a new `instructions_format` dict to the `ToolSpec` class.

Again, before going further let me know what you think.

Good news: now that ollama supports the tools in stream mode, I was able to use basics tools with  the small `llama3.2:1B` model.

File: gptme/tools/base.py
Comment (jrmi) [reply gptme/tools/base.py]: Good idea. Done.

File: gptme/tools/base.py
Comment (jrmi) [reply gptme/tools/base.py]: Done.

File: gptme/tools/base.py
Comment (jrmi) [reply gptme/tools/base.py]: I wasn't happy about the naming. Removed now.

File: gptme/tools/save.py
Comment (jrmi) [reply gptme/tools/save.py]: Did it.

File: gptme/tools/base.py
Comment (jrmi) [reply gptme/tools/base.py]: I added a compat mode that should work for most of the tool like you suggested but I believe that we should migrate them all we proper argument when it's possible and it makes sense because it helps the model to use the tool properly.

File: gptme/tools/base.py
Comment (jrmi) [reply gptme/tools/base.py]: Removed.

File: gptme/tools/base.py
Comment (ErikBjare) [inline gptme/tools/base.py]:
```suggestion
    def to_output(self, tool_format: ToolFormat = "markdown") -> str:
```

File: gptme/tools/save.py
Comment (ErikBjare) [inline gptme/tools/save.py]: I prefer the shorter `path` (many occurrences)

```suggestion
            name="path",
```

File: gptme/tools/shell.py
Comment (ErikBjare) [inline gptme/tools/shell.py]: This isn't markdown-specific, unfortunately.

File: gptme/tools/base.py
Comment (ErikBjare) [inline gptme/tools/base.py]:
```suggestion
    examples: str | Callable[[str], str] = ""
```

File: gptme/tools/base.py
Comment (ErikBjare) [inline gptme/tools/base.py]: Assuming both are stripped strings:

```suggestion
            return "\n\n".join([self.instructions, self.instructions_format[tool_format]])
```

File: gptme/tools/save.py
Comment (ErikBjare) [inline gptme/tools/save.py]: Will this later be written like this?

```suggestion
{ToolUse("save", {"path": "hello.py", "content": 'print("Hello world")'}).to_output(tool_type)}
```

Or how can we eliminate the duplicated args?

File: pyproject.toml
Comment (ErikBjare) [inline pyproject.toml]: Place this line in an appropriate section (core)

**Review by ErikBjare — COMMENTED** [review]
Very nice! This looking really good.

Just a few comments.

File: gptme/tools/base.py
Comment (jrmi) [reply gptme/tools/base.py]: Fixed.

File: gptme/tools/save.py
Comment (jrmi) [reply gptme/tools/save.py]: Fixed.

File: gptme/tools/shell.py
Comment (jrmi) [reply gptme/tools/shell.py]: Fixed.

File: gptme/tools/base.py
Comment (jrmi) [reply gptme/tools/base.py]: Fixed

File: gptme/tools/base.py
Comment (jrmi) [reply gptme/tools/base.py]: Fixed

File: gptme/tools/save.py
Comment (jrmi) [reply gptme/tools/save.py]: Fixed.

File: pyproject.toml
Comment (jrmi) [reply pyproject.toml]: Fixed.

**Comment by jrmi** [comment]
> Very nice! This looking really good.
> 
> Just a few comments.

Thanks for your second review. I think I fixed all your feedbacks and I have adapted all the tools. 

I faced a new issue: the python tool description is too long (>1024 char limit of tool description) because we include the docstring of the python function. I imagined two solutions:
- we could make the description of the `chats` tool shorter as it's the longest one but we'll encounter this issue sooner or later down the road.
- We move the functions description in each tool instead.

For now I have implemented the second version but I could try the other if you prefer.

Ready for another review.

File: gptme/tools/base.py
Comment (jrmi) [reply gptme/tools/base.py:294]: `json_repair` documentation says we [shouldn't try to load the json](https://github.com/mangiucugna/json_repair/?tab=readme-ov-file#avoid-this-antipattern) ourselves and we should use it as a drop in replacement of `json.loads`.

**Review by ErikBjare — APPROVED** [review]
Looks great!

I only have nitpick comments that I think will be easier to fix for me after merge (incl. organizing `openai` and `anthropic` imports so that they are lazy-loaded, to improve startup times).

I have yet to try running it, does it feel solid to you? Do the tests pass locally?

If the markdown/default format still works reliably, happy to merge! I can resolve the merge conflict via rebase (caused by merging https://github.com/ErikBjare/gptme/pull/293).

**Comment by jrmi** [comment]
> I have yet to try running it, does it feel solid to you? Do the tests pass locally?

The tests pass locally and I have added more tests. The tool format probably need a bit of tweaking but it's an opt-in option and AFAIK the default markdown version doesn't seem to be affected by the changes. so I would say it's safe.

> If the markdown/default format still works reliably, happy to merge! I can resolve the merge conflict via rebase (caused by merging https://github.com/ErikBjare/gptme/pull/293).

The only doubt is that I can't test the Anthropic version because I've no API key. May be worth to try at least one prompt?

**Comment by ErikBjare** [comment]
@jrmi I just sent you an Anthropic API key in DM on Discord.

Will try to get this merged sometime today.

**Comment by ErikBjare** [comment]
Merged in #300, thanks a lot for contributing!

Tracking remaining issues in https://github.com/ErikBjare/gptme/issues/302

Made a PR with some fixes: https://github.com/ErikBjare/gptme/pull/303

**Pull request closed without merging**
